### PR TITLE
kvserver: refactor replicate queue to enable allocator code reuse

### PR DESCRIPTION
**Pull request by AlexTalks**
This change refactors parts of the replicate queue's `PlanOneChange(..)`
and `addOrRemove{Non}Voters(..)` functions to reusable helper functions
that simplify usage of the allocator and deduplicate repeated code
paths. The change also adds convenience methods to the `AllocatorAction`
enum, to move certain determinations (such as if a computed allocator
action is a remove or a replace) closer to the allocator type it is
based on. These changes further enable the ability to use the allocator
outside of the replicate queue, and enable the ability for some of this
logic to move into the allocator itself in the future.

Depends on #91941.

Epic: none

Release note: None

### CODE REVIEW

File: pkg/kv/kvserver/replicate_queue.go
Comment (kvoli) [inline pkg/kv/kvserver/replicate_queue.go:1178]: We should certainly  try  and  move this into the `allocatorimpl` pkg.

**Review by AlexTalks — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/94114)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @kvoli)

___
*[`pkg/kv/kvserver/replicate_queue.go` line 1178 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/94114#-NJr2YhrJXNt8pJFRjjK-r1-1178:-NJvqTaT7ULS_ZfiBDPS:b-fajfam) ([raw file](https://github.com/cockroachdb/cockroach/blob/fdce47e413dbc66d1ca05d1d6c61f083e1a4afa1/pkg/kv/kvserver/replicate_queue.go#L1178)):*
<details><summary><i>Previously, kvoli (Austen) wrote…</i></summary><blockquote>

We should certainly  try  and  move this into the `allocatorimpl` pkg. 
</blockquote></details>

Done.

Also moved the "avoid fragile quorum" check into the allocator as well.

**Review by kvoli — COMMENTED** [review]
Looks good. There is one test failing that I would look into:

TestPromoteNonVoterInAddVoter

There is fragile logic which decides to promote a non-voter on removal.

> Reviewed 1 of 2 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/94114#-:-NKJBRVtDpR6Jy4aJ3lK:bepuy2c)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @AlexTalks)

___
*[`pkg/kv/kvserver/replicate_queue.go` line 1303 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/94114#-NKJAtJA7psEVxClXRmj:-NKJAtJA7psEVxClXRmk:bu509yg) ([raw file](https://github.com/cockroachdb/cockroach/blob/c62636b72a54ebf846c2450df9a59be3c4ab2034/pkg/kv/kvserver/replicate_queue.go#L1303)):*
> ```Go
> ) (op AllocationOp, _ error) {
> 	effects := effectBuilder{}
> 	conf := repl.SpanConfig()
> ```

This should be passed in to be consistent with the other commit that makes desc/conf consistent.

**Review by AlexTalks — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/94114)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @kvoli)

___
*[`pkg/kv/kvserver/replicate_queue.go` line 1303 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/94114#-NKJAtJA7psEVxClXRmj:-NL2kPI0CgyX22Hr6CHQ:b-x4h0cr) ([raw file](https://github.com/cockroachdb/cockroach/blob/c62636b72a54ebf846c2450df9a59be3c4ab2034/pkg/kv/kvserver/replicate_queue.go#L1303)):*
<details><summary><i>Previously, kvoli (Austen) wrote…</i></summary><blockquote>

This should be passed in to be consistent with the other commit that makes desc/conf consistent. 
</blockquote></details>

Some work will be required to rebase that change on top of all the other changes to the replicate queue/allocator, so I'll wait until these are merged to do that.

**Review by kvoli — APPROVED** [review]
:lgtm:

> Reviewed 2 of 2 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/94114#-:-NL8eq7PBTBCJ9Oapx2t:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @AlexTalks)

___
*[`pkg/kv/kvserver/replicate_queue.go` line 1178 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/94114#-NJr2YhrJXNt8pJFRjjK-r1-1178:-NL8eSNe6O84zkOvh7u1:btojbfd) ([raw file](https://github.com/cockroachdb/cockroach/blob/fdce47e413dbc66d1ca05d1d6c61f083e1a4afa1/pkg/kv/kvserver/replicate_queue.go#L1178)):*
<details><summary><i>Previously, AlexTalks (Alex Sarkesian) wrote…</i></summary><blockquote>

Done.

Also moved the "avoid fragile quorum" check into the allocator as well.
</blockquote></details>

Great!
___
*[`pkg/kv/kvserver/replicate_queue.go` line 1303 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/94114#-NKJAtJA7psEVxClXRmj:-NL8e4R42vG6BdAGk1DR:b2nhysk) ([raw file](https://github.com/cockroachdb/cockroach/blob/c62636b72a54ebf846c2450df9a59be3c4ab2034/pkg/kv/kvserver/replicate_queue.go#L1303)):*
<details><summary><i>Previously, AlexTalks (Alex Sarkesian) wrote…</i></summary><blockquote>

Some work will be required to rebase that change on top of all the other changes to the replicate queue/allocator, so I'll wait until these are merged to do that.
</blockquote></details>

ack
___
*[`pkg/kv/kvserver/replicate_queue.go` line 1038 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/94114#-NL8elzX1wwUZeqxdADY:-NL8elzX1wwUZeqxdADZ:btni5mr) ([raw file](https://github.com/cockroachdb/cockroach/blob/f35e4fc4fbd398a99858fb7a982cf6393dcce2c0/pkg/kv/kvserver/replicate_queue.go#L1038)):*
> ```Go
> 				action,
> 				voterReplicas, nonVoterReplicas,
> 				liveVoterReplicas, deadVoterReplicas,
> ```

Just curious - was this the source of the test failures previously?
___
*[`pkg/kv/kvserver/allocator/allocatorimpl/allocator.go` line 136 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/94114#-NL8e7md8nrTUqL_jyBO:-NL8e7md8nrTUqL_jyBP:b-tpfh90) ([raw file](https://github.com/cockroachdb/cockroach/blob/f35e4fc4fbd398a99858fb7a982cf6393dcce2c0/pkg/kv/kvserver/allocator/allocatorimpl/allocator.go#L136)):*
> ```Go
> )
> 
> func (a AllocatorAction) Add() bool {
> ```

nit: does the linter enforce export function commenting? These are pretty self explanatory, just wondering why that got removed.

**Review by AlexTalks — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/94114)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @kvoli)

___
*[`pkg/kv/kvserver/replicate_queue.go` line 1303 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/94114#-NKJAtJA7psEVxClXRmj:-NL92hg1AzUUJzBtSWLv:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/c62636b72a54ebf846c2450df9a59be3c4ab2034/pkg/kv/kvserver/replicate_queue.go#L1303)):*
<details><summary><i>Previously, kvoli (Austen) wrote…</i></summary><blockquote>

ack
</blockquote></details>

Done.
___
*[`pkg/kv/kvserver/replicate_queue.go` line 1038 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/94114#-NL8elzX1wwUZeqxdADY:-NL92XZV9U38z58sGrcV:b-5fhm6y) ([raw file](https://github.com/cockroachdb/cockroach/blob/f35e4fc4fbd398a99858fb7a982cf6393dcce2c0/pkg/kv/kvserver/replicate_queue.go#L1038)):*
<details><summary><i>Previously, kvoli (Austen) wrote…</i></summary><blockquote>

Just curious - was this the source of the test failures previously?
</blockquote></details>

One of them, unfortunately.
___
*[`pkg/kv/kvserver/allocator/allocatorimpl/allocator.go` line 136 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/94114#-NL8e7md8nrTUqL_jyBO:-NL92kdZ9LDPB-k57DeS:b-ng64w5) ([raw file](https://github.com/cockroachdb/cockroach/blob/f35e4fc4fbd398a99858fb7a982cf6393dcce2c0/pkg/kv/kvserver/allocator/allocatorimpl/allocator.go#L136)):*
<details><summary><i>Previously, kvoli (Austen) wrote…</i></summary><blockquote>

nit: does the linter enforce export function commenting? These are pretty self explanatory, just wondering why that got removed.
</blockquote></details>

I don't think so, can add if needed.

**Pull request merged**
